internal/cluster/watch.go

```diff
@@ -0,0 +1,190 @@
+func WatchPipelineActivity(namespace string, jxClient *jenkinsclientv1.JenkinsV1Client) {
+	listWatch := cache.NewListWatchFromClient(jxClient.RESTClient(), "pipelineactivities", namespace, fields.Everything())
+	kube.SortListWatchByName(listWatch)
+	_, actController := cache.NewInformer(
+		listWatch,
+		&jenkinsv1.PipelineActivity{},
+		time.Minute*10,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				onPipelineActivity(obj, jxClient)
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				onPipelineActivity(newObj, jxClient)
+			},
+			DeleteFunc: func(obj interface{}) {},
+		},
+	)
+	stop := make(chan struct{})
+	go actController.Run(stop)
+}
+
+func onPipelineActivity(obj interface{}, jxClient *jenkinsclientv1.JenkinsV1Client) {
+	pipelineActivity, ok := obj.(*jenkinsv1.PipelineActivity)
+	if !ok {
+		logger.Warnf("unexpected type %T", obj)
+	} else {
+		err := handlePipelineActivity(pipelineActivity, jxClient)
+		if err != nil {
+			logger.Error(err)
+		}
+	}
+}
+
+func handlePipelineActivity(pipelineActivity *jenkinsv1.PipelineActivity, jxClient *jenkinsclientv1.JenkinsV1Client) (err error) {
+	for _, attachment := range pipelineActivity.Spec.Attachments {
+		if attachment.Name != "jacoco" {
+			continue
+		}
+
+		// TODO Handle having multiple attachments properly
+		for _, url := range attachment.URLs {
+			url = fmt.Sprintf("%s?version=%d", url, time.Now().UnixNano()/int64(time.Millisecond))
+			report, err := report.RetrieveReport(url)
+			if err != nil {
+				logger.Errorf("unable to retrieve %s for processing: %s", url, err)
+				continue
+			}
+
+			fact := createFact(report, url)
+			updatePipelineActivity(pipelineActivity.Name, pipelineActivity.Namespace, fact, jxClient)
+		}
+	}
+	return nil
+}
+
+func updatePipelineActivity(name string, namespace string, fact jenkinsv1.Fact, jxClient *jenkinsclientv1.JenkinsV1Client) {
+	var pipelineActivity *jenkinsv1.PipelineActivity
+	var err error
+	// re-get the pipeline activity
+	f := func() error {
+		pipelineActivity, err = jxClient.PipelineActivities(namespace).Get(name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+
+	err = util.ApplyWithBackoff(f)
+	if err != nil {
+		logger.Errorf("error retrieving PipelineActivity %s with uuid %s for update: %s", pipelineActivity.Name, pipelineActivity.UID, err)
+		return
+	}
+
+	found, index, err := indexOfJacocoCoverageFactType(pipelineActivity)
+	if err != nil {
+		logger.Error(err)
+	}
+
+	if found {
+		// Updating the existing jacoco report with the new one. Is this what we want? (HF)
+		pipelineActivity.Spec.Facts[index] = fact
+	} else {
+		pipelineActivity.Spec.Facts = append(pipelineActivity.Spec.Facts, fact)
+	}
+
+	// update the pipeline CRD
+	f = func() error {
+		_, err = jxClient.PipelineActivities(pipelineActivity.Namespace).Update(pipelineActivity)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+	err = util.ApplyWithBackoff(f)
```

Comment:
Here is the exponential re-try for updating the CRD

Reply:
So this will not help with 409s, since what you would need to retry is basically the whole `updatePipelineActivity` method: getting any existing PA, appending to or editing its facts, and updating it.

Reply:
You are right. Let me take care of this.